File: C#/Memento.cs
using System;
using System.Collections.Generic;

class Originator {
    private string state;

    public void SetState(string state) {
        this.state = state;
    }

    public string GetState() {
        return state;
    }

    public Memento Save() {
        return new Memento(state);
    }

    public void Restore(Memento memento) {
        state = memento.GetState();
    }
}

class Memento {
    private readonly string state;

    public Memento(string state) {
        this.state = state;
    }

    public string GetState() {
        return state;
    }
}

class Caretaker {
    private readonly List<Memento> mementos = new List<Memento>();
    private readonly Originator originator;

    public Caretaker(Originator originator) {
        this.originator = originator;
    }

    public void Save() {
        mementos.Add(originator.Save());
    }

    public void Restore(int index) {
        originator.Restore(mementos[index]);
    }
}

class Program {
    static void Main(string[] args) {
        Originator originator = new Originator();
        Caretaker caretaker = new Caretaker(originator);

        originator.SetState("State 1");
        caretaker.Save();

        originator.SetState("State 2");
        caretaker.Save();

        originator.SetState("State 3");

        Console.WriteLine("Current state: " + originator.GetState());

        caretaker.Restore(1);

        Console.WriteLine("Restored state: " + originator.GetState());
    }
}

File: Dart/Memento.dart
class Memento {
  final String _state;
  Memento(this._state);
  String get state => _state;
}

class Originator {
  String _state;
  Originator(this._state);
  Memento createMemento() => Memento(_state);
  void restoreMemento(Memento memento) {
    _state = memento.state;
  }
  String get state => _state;
  set state(String state) {
    _state = state;
  }
}

class Caretaker {
  final List<Memento> _mementos = [];
  void addMemento(Memento memento) {
    _mementos.add(memento);
  }
  Memento getMemento(int index) => _mementos[index];
}

void main() {
  final originator = Originator('State1');
  final caretaker = Caretaker();

  caretaker.addMemento(originator.createMemento());
  originator.state = 'State2';

  caretaker.addMemento(originator.createMemento());
  originator.state = 'State3';

  print('Current state: ${originator.state}');
  originator.restoreMemento(caretaker.getMemento(1));
  print('Restored state: ${originator.state}');
}
